### PR TITLE
feat(HostnameGenerator): apply templates to MeshServices

### PR DESCRIPTION
**Pull request by michaelbeaumont**
- [x] interval config

### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues -- https://github.com/kumahq/kuma/issues/10034
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [x] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/dns/hostname_generator.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	if !slices.Contains(rt.Config().CoreResources.Enabled, "hostnamegenerators") {
+		logger.Info("HostnameGenerator is not enabled, not starting generator")
+		return nil
+	}
+	generator, err := hostname.NewGenerator(
+		logger,
+		rt.Metrics(),
+		rt.ResourceManager(),
+		100*time.Millisecond,
+		// rt.Config().IPAM.MeshService.AllocationInterval.Duration, // TODO
```
Comment (lukidzi) [inline pkg/dns/hostname_generator.go:28]: Question: Should this interval be the same for different resources? `MeshExternalService`needs hostname generator and I think the only needed thing is to add support [here](https://github.com/kumahq/kuma/pull/10362/files#diff-4de19adee6a0c1d08e3f861fc716e75a2b540a5809e65b0a840659c1cf96b996) so it seems that we don't need a separate component for a MeshExternalService

**Review by lukidzi — APPROVED** [review]

File: pkg/dns/hostname_generator.go (hunk 1 above)
Comment (michaelbeaumont) [reply pkg/dns/hostname_generator.go:28]: No you're right, we should reuse it and we shouldn't need separate configs per resource type.

File: pkg/core/resources/apis/meshservice/api/v1alpha1/meshservice.go [hunk 2]
```diff
@@ -53,8 +63,70 @@ type TLS struct {
...
+const (
+	GeneratedCondition string = "Generated"
+)
+
+const (
+	GeneratedReason     string = "Generated"
+	TemplateErrorReason string = "TemplateError"
+)
+
+type Condition struct {
```
Comment (lukidzi) [inline pkg/core/resources/apis/meshservice/api/v1alpha1/meshservice.go:79]: Should this be a common object? I think `MeshExternalService` could have the same structure

File: pkg/dns/hostname_generator.go (hunk 1 above)
Comment (slonka) [reply pkg/dns/hostname_generator.go:28]: Why would we need to pass the interval of IPAM to HostnameGenerator? Do they need to run in sync?

File: pkg/core/resources/apis/hostnamegenerator/hostname/generator.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+			"label": func(key string) (string, error) {
+				val, ok := service.GetMeta().GetLabels()[key]
+				if !ok {
+					return "", errors.Errorf("label %s not found", key)
+				}
+				return val, nil
+			},
+		},
+	)
+	tmpl, err := tmpl.Parse(generator.Spec.Template)
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/apis/hostnamegenerator/hostname/generator.go:94]: Do we validate this statically on resource apply as well?

File: pkg/dns/hostname_generator.go (hunk 1 above)
Comment (michaelbeaumont) [reply pkg/dns/hostname_generator.go:28]: No, we just need _some_ interval

File: pkg/core/resources/apis/hostnamegenerator/hostname/generator.go (hunk 3 above)
Comment (michaelbeaumont) [reply pkg/core/resources/apis/hostnamegenerator/hostname/generator.go:94]: added

File: pkg/core/resources/apis/meshservice/api/v1alpha1/meshservice.go (hunk 2 above)
Comment (michaelbeaumont) [reply pkg/core/resources/apis/meshservice/api/v1alpha1/meshservice.go:79]: I'll leave that to the PR that adds MeshExternalService

File: pkg/core/resources/apis/hostnamegenerator/hostname/generator.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+			return errors.Wrap(err, "couldn't update MeshService status")
+		}
+	}
+
+	return nil
+}
+
+func (g *Generator) NeedLeaderElection() bool {
+	return true
+}
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/apis/hostnamegenerator/hostname/generator.go:224]: do we have conflict resolution here somewhere?

File: pkg/core/resources/apis/hostnamegenerator/hostname/generator.go [hunk 5]
```diff
@@ -155,8 +178,16 @@ func (g *Generator) generateHostnames(ctx context.Context) error {
 					message = err.Error()
 				}
 				if generated != "" {
-					generationConditionStatus = kube_meta.ConditionTrue
-					reason = meshservice_api.GeneratedReason
+					if key, ok := generatedHostnames[generated]; ok && key != serviceKey {
+						generationConditionStatus = kube_meta.ConditionFalse
+						reason = meshservice_api.CollisionReason
+						message = "Hostname collision"
```
Comment (jakubdyszkiewicz) [inline pkg/core/resources/apis/hostnamegenerator/hostname/generator.go:184]: can we add "Hostname collision with service XYZ"?

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**